### ISSUE TITLE
Prepare for Python Markdown inline processor changes

### ISSUE THREAD

**Issue by facelessuser** [facelessuser/pymdown-extensions#215]
I am currently reworking inline processor in a new InlineProcessor class for Python Markdown which improves inline performance, but it also adds the ability to do more analysis and alter the bounds of the original match (ref https://github.com/Python-Markdown/markdown/issues/551).  This is much more flexible. For example, we can match just `[` for links, and then in the plugin tokenize the next characters handling recursive `[]` which Python's re cannot handle.  Python Markdown is moving all inline processors in that direction and providing the legacy versions of the old methods for the basic patterns, but there are two `InlineProcessors` we use that are not going to be accessible after the change `LinkPattern` and `HtmlPattern`.  So we need to simply prepare by checking now whether we can import `LinkPattern` in MagicLinks, and `HtmlPattern` in SmartSymbols.

We will provide alternate logic for these two cases (import succeeds and import fails).  All other patterns can use legacy logic for now. After the Python Markdown finally releases the change, we can change all of pymdown-extensions to utilize the new `InlineProcessor` classes and drop the `Pattern` classes.  Then we can lock pymdown-extensions to whatever version of Markdown that introduces the new `InlineProcessor` classes.

So basically there are two phases:

- [x] Compatibility phase -- do the bare minimum to keep working (make sure we work now, and once the changes are released)
- [ ] Full implementation phase (we fully use the new InlineProcessor class and abandon Pattern).

**Comment by facelessuser** [comment]
This will be handled under #351.
